File: backend/src/data/ensureVectorIndex.ts
import type { PrismaClient } from '@prisma/client'
import { createEmbeddings } from '../ai/embeddings'
import { SqliteVecStore } from '../ai/vectorStore'
import { flattenKnowledge } from './clubKnowledge'

type EnsureOptions = {
  /** 为 true 时即使表非空也重建（脚本用）。 */
  force?: boolean
}

/**
 * 构建/重建社团知识的向量索引。
 * 仿 `ensureDemoData`：无 Key 或已建好则跳过（除非 force）。
 * @returns 写入的知识块数量（跳过时为 0）
 */
export async function ensureVectorIndex(
  prisma: PrismaClient,
  options: { enabled: boolean; apiKey: string } & EnsureOptions,
): Promise<number> {
  if (!options.enabled || !options.apiKey) return 0

  const store = new SqliteVecStore()
  try {
    if (!options.force && !store.isEmpty()) return 0

    const clubs = await prisma.club.findMany({ select: { id: true, name: true } })
    const clubsByName = new Map(clubs.map(club => [club.name, club.id]))
    const docs = flattenKnowledge(clubsByName)
    if (docs.length === 0) return 0

    const embeddings = createEmbeddings()
    const vectors = await embeddings.embedDocuments(docs.map(doc => doc.content))
    store.indexDocuments(docs, vectors)
    return docs.length
  } finally {
    store.close()
  }
}

File: backend/app/data/ensure_vector_index.py
"""构建/重建社团知识的向量索引。等价于原 TS data/ensureVectorIndex.ts。
无 Key 或已建好则跳过（除非 force）。返回写入的知识块数量。
"""
from __future__ import annotations

from sqlalchemy import select

from ..ai.embeddings import create_embeddings
from ..ai.vector_store import ClubDoc, SqliteVecStore
from ..db import Club, SessionLocal
from .club_knowledge import flatten_knowledge


async def ensure_vector_index(enabled: bool, api_key: str, force: bool = False) -> int:
    if not enabled or not api_key:
        return 0

    store = SqliteVecStore()
    try:
        if not force and not store.is_empty():
            return 0

        with SessionLocal() as session:
            clubs = session.execute(select(Club.id, Club.name)).all()
        clubs_by_name = {name: club_id for club_id, name in clubs}
        docs = [ClubDoc(**d) for d in flatten_knowledge(clubs_by_name)]
        if not docs:
            return 0

        embeddings = create_embeddings()
        vectors = await embeddings.aembed_documents([doc.content for doc in docs])
        store.index_documents(docs, vectors)
        return len(docs)
    finally:
        store.close()
